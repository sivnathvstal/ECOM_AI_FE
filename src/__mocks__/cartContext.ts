import { Cart, CartProduct } from '../types';

export const mockCartItems: CartProduct[] = [
  {
    productId: 1,
    quantity: 2,
    product: {
      id: 1,
      title: 'Test Product 1',
      price: 10.99,
      description: 'Test description 1',
      category: 'electronics',
      image: 'test1.jpg',
      rating: { rate: 4.5, count: 100 },
    },
  },
  {
    productId: 2,
    quantity: 1,
    product: {
      id: 2,
      title: 'Test Product 2',
      price: 20.99,
      description: 'Test description 2',
      category: 'clothing',
      image: 'test2.jpg',
      rating: { rate: 4.0, count: 80 },
    },
  },
];

export const mockCart: Cart = {
  id: 1,
  userId: 1,
  date: '2023-01-01',
  products: [
    { productId: 1, quantity: 2 },
    { productId: 2, quantity: 1 },
  ],
};

export const mockCartContext = {
  cart: mockCart,
  cartItems: mockCartItems,
  isLoading: false,
  error: null,
  addItem: jest.fn(),
  updateItem: jest.fn(),
  removeItem: jest.fn(),
  totalItems: 3, // 2 + 1
  totalPrice: 42.97, // 10.99*2 + 20.99*1
};

export const mockEmptyCartContext = {
  cart: {
    id: 0,
    userId: 1,
    date: '2023-01-01',
    products: [],
  },
  cartItems: [],
  isLoading: false,
  error: null,
  addItem: jest.fn(),
  updateItem: jest.fn(),
  removeItem: jest.fn(),
  totalItems: 0,
  totalPrice: 0,
};

export const mockLoadingCartContext = {
  ...mockCartContext,
  isLoading: true,
};

export const mockErrorCartContext = {
  ...mockCartContext,
  error: 'Failed to fetch cart',
};

const defaultExport = {
  mockCartItems,
  mockCart,
  mockCartContext,
  mockEmptyCartContext,
  mockLoadingCartContext,
  mockErrorCartContext,
};
export default defaultExport; 