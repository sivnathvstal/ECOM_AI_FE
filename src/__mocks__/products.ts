import { Product, PaginationParams, ProductsResponse } from '../types';

export const mockProducts: Product[] = [
  {
    id: 1,
    title: 'Test Product 1',
    price: 10.99,
    description: 'Test description 1',
    category: 'electronics',
    image: 'test1.jpg',
    rating: { rate: 4.5, count: 100 },
  },
  {
    id: 2,
    title: 'Test Product 2',
    price: 20.99,
    description: 'Test description 2',
    category: 'clothing',
    image: 'test2.jpg',
    rating: { rate: 4.0, count: 80 },
  },
  {
    id: 3,
    title: 'Test Product 3',
    price: 15.99,
    description: 'Test description 3',
    category: 'electronics',
    image: 'test3.jpg',
    rating: { rate: 3.5, count: 60 },
  },
  {
    id: 4,
    title: 'Test Product 4',
    price: 25.99,
    description: 'Test description 4',
    category: 'jewelry',
    image: 'test4.jpg',
    rating: { rate: 4.8, count: 120 },
  },
];

export const mockPagination: PaginationParams = {
  page: 1,
  limit: 8,
  total: 4,
};

export const mockProductsResponse: ProductsResponse = {
  products: mockProducts,
  pagination: mockPagination,
};

export const mockCategories: string[] = [
  'electronics',
  'clothing',
  'jewelry',
]; 