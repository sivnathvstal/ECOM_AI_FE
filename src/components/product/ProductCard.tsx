import React from 'react';
import {
  Card,
  CardActionArea,
  CardActions,
  CardContent,
  CardMedia,
  Button,
  Typography
} from '@material-ui/core';
import { Link as RouterLink } from 'react-router-dom';
import { Product } from '../../types';
import './ProductCard.css';

interface ProductCardProps {
  product: Product;
}

const ProductCard: React.FC<ProductCardProps> = ({ product }) => {
  const { id, title, price, image } = product;

  return (
    <Card className="product-card" data-testid={`product-card-${id}`}>
      <CardActionArea component={RouterLink} to={`/product/${id}`}>
        <CardMedia
          className="product-media"
          image={image}
          title={title}
          data-testid={`product-image-${id}`}
        />
        <CardContent className="product-content">
          <Typography 
            gutterBottom 
            variant="h6" 
            component="h2" 
            className="product-title"
            data-testid={`product-title-${id}`}
          >
            {title}
          </Typography>
          <Typography 
            variant="body1" 
            color="textPrimary" 
            component="p" 
            className="product-price"
            data-testid={`product-price-${id}`}
          >
            ${price.toFixed(2)}
          </Typography>
        </CardContent>
      </CardActionArea>
      <CardActions className="product-actions">
        <Button 
          size="small" 
          color="primary" 
          component={RouterLink} 
          to={`/product/${id}`}
          data-testid={`view-details-${id}`}
        >
          View Details
        </Button>
      </CardActions>
    </Card>
  );
};

export default ProductCard; 