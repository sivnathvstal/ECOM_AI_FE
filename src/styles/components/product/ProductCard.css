.product-card {
  max-width: 345px;
  height: 100%;
  display: flex;
  flex-direction: column;
}

.product-media {
  height: 200px;
  background-size: contain;
  margin: 8px;
}

.product-content {
  flex-grow: 1;
}

.product-title {
  overflow: hidden;
  text-overflow: ellipsis;
  display: -webkit-box;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
  height: 3em;
}

.product-price {
  font-weight: bold;
  color: #3f51b5; /* primary color */
  margin-top: 8px;
}

.product-actions {
  justify-content: space-between;
} 