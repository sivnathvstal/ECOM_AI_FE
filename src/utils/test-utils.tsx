import React, { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { ThemeProvider, createTheme } from '@material-ui/core';
import { QueryClient, QueryClientProvider } from 'react-query';
import { CartProvider } from '../context/CartContext';

// Create a theme for testing
const theme = createTheme({
  palette: {
    primary: {
      main: '#3f51b5',
    },
    secondary: {
      main: '#f50057',
    },
  },
});

// Create a client for React Query
const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
    },
  },
});

// All providers wrapper
const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <QueryClientProvider client={queryClient}>
      <ThemeProvider theme={theme}>
        <CartProvider>
          <BrowserRouter>{children}</BrowserRouter>
        </CartProvider>
      </ThemeProvider>
    </QueryClientProvider>
  );
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, ...options });

// Re-export everything
export * from '@testing-library/react';

// Override render method
export { customRender as render }; 