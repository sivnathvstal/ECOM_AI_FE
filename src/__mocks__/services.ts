import { mockProducts, mockProductsResponse, mockCategories } from './products';
import { mockCart, mockCartItems } from './cartContext';

// Product service mocks
export const mockProductService = {
  getProducts: jest.fn().mockResolvedValue(mockProductsResponse),
  getProductById: jest.fn().mockImplementation((id: number) => {
    const product = mockProducts.find(p => p.id === id);
    if (product) {
      return Promise.resolve(product);
    }
    return Promise.reject(new Error('Product not found'));
  }),
  getCategories: jest.fn().mockResolvedValue(mockCategories),
  searchProducts: jest.fn().mockResolvedValue(mockProductsResponse),
};

// Cart service mocks
export const mockCartService = {
  getCart: jest.fn().mockResolvedValue(mockCart),
  addToCart: jest.fn().mockResolvedValue(mockCart),
  updateCartItem: jest.fn().mockResolvedValue(mockCart),
  removeFromCart: jest.fn().mockResolvedValue(mockCart),
  getCartWithProducts: jest.fn().mockResolvedValue(mockCartItems),
}; 