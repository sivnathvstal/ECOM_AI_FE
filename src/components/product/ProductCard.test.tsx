import React from 'react';
import { render, screen } from '../../utils/test-utils';
import ProductCard from './ProductCard';
import { mockProducts } from '../../__mocks__/products';

describe('ProductCard', () => {
  const product = mockProducts[0];

  it('renders correctly', () => {
    const { container } = render(<ProductCard product={product} />);
    expect(container).toMatchSnapshot();
  });

  it('displays product information correctly', () => {
    render(<ProductCard product={product} />);
    
    expect(screen.getByTestId(`product-title-${product.id}`)).toHaveTextContent(product.title);
    expect(screen.getByTestId(`product-price-${product.id}`)).toHaveTextContent(`$${product.price.toFixed(2)}`);
    
    // Check if the image is rendered with correct attributes
    const productImage = screen.getByTestId(`product-image-${product.id}`);
    expect(productImage).toHaveAttribute('title', product.title);
  });

  it('has correct link to product details', () => {
    render(<ProductCard product={product} />);
    
    // Check if the view details button links to the correct product page
    const viewDetailsButton = screen.getByTestId(`view-details-${product.id}`);
    expect(viewDetailsButton).toHaveTextContent('View Details');
    expect(viewDetailsButton).toHaveAttribute('href', `/product/${product.id}`);
  });
}); 