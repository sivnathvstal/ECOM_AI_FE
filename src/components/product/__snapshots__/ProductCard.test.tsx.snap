// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard renders correctly 1`] = `
<div>
  <div
    class="MuiPaper-root MuiCard-root product-card MuiPaper-elevation1 MuiPaper-rounded"
    data-testid="product-card-1"
  >
    <a
      aria-disabled="false"
      class="MuiButtonBase-root MuiCardActionArea-root"
      href="/product/1"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiCardMedia-root product-media"
        data-testid="product-image-1"
        style="background-image: url(test1.jpg);"
        title="Test Product 1"
      />
      <div
        class="MuiCardContent-root product-content"
      >
        <h2
          class="MuiTypography-root product-title MuiTypography-h6 MuiTypography-gutterBottom"
          data-testid="product-title-1"
        >
          Test Product 1
        </h2>
        <p
          class="MuiTypography-root product-price MuiTypography-body1 MuiTypography-colorTextPrimary"
          data-testid="product-price-1"
        >
          $
          10.99
        </p>
      </div>
      <span
        class="MuiCardActionArea-focusHighlight"
      />
      <span
        class="MuiTouchRipple-root"
      />
    </a>
    <div
      class="MuiCardActions-root product-actions MuiCardActions-spacing"
    >
      <a
        aria-disabled="false"
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall"
        data-testid="view-details-1"
        href="/product/1"
        role="button"
        tabindex="0"
      >
        <span
          class="MuiButton-label"
        >
          View Details
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </a>
    </div>
  </div>
</div>
`;
